Rewrite `parse_address` util and add one test

hramezani opened the pull request:
Maybe better implementation of `parse_address` util.

hramezani commented:
@berkerpeksag  thanks,
I think it is a cleaner and easy to underestanding implementation.
I think there is no improvement in speed and performance

hramezani commented:
@berkerpeksag, I changed the port conversation to int.

tilgovi commented on gunicorn/util.py:
```diff
@@ -232,33 +232,29 @@ def is_ipv6(addr):
 
-def parse_address(netloc, default_port=8000):
+def parse_address(netloc, default_port='8000'):
+    host, port = [netloc, default_port]
```
I would move this down to an `else` statement. That would also make sure the `tcp://` stripping take effect. Right now, `tcp://` with an address but no port might not work.

tilgovi replied:
I would move this down to an `else` statement or at least after these first two blocks. Probably an `else` because I like branching where all assignments are at the same level/depth. ;) That would also make sure the `tcp://` stripping take effect. Right now, `tcp://` with an address but no port might not work.

tilgovi commented on gunicorn/util.py:
```diff
@@ -232,33 +232,29 @@ def is_ipv6(addr):
     if '[' in netloc and ']' in netloc:
-        host = netloc.split(']')[0][1:].lower()
+        host = netloc.split(']')[0][1:]
+        port = (netloc.split(']:') + [default_port])[-1]
```
I feel like this should be `[1]` instead of `[-1]`? We want _either_ the port that was specified in the netloc or the default port, right?

If we're rewriting this code I think we can make it less tricky by being a bit clearer about this stuff.

hramezani commented:
@tilgovi, thanks for your comments, I changed my code as your request and add some test for covering `tpc://` address

hramezani commented:
@tilgovi @benoitc any news?

berkerpeksag commented on gunicorn/util.py:
```diff
@@ -232,33 +232,30 @@ def is_ipv6(addr):
 
-def parse_address(netloc, default_port=8000):
+def parse_address(netloc, default_port='8000'):
+
```
Style nit: No need to add a newline.

berkerpeksag commented on gunicorn/util.py:
```diff
@@ -232,33 +232,30 @@ def is_ipv6(addr):
         netloc = netloc.split("tcp://")[1]
 
-    # get host
+    host, port = [netloc, default_port]
```
No need to create a new list here. `host, port = netloc, default_port` should work, right?

berkerpeksag commented on gunicorn/util.py:
```diff
@@ -232,33 +232,30 @@ def is_ipv6(addr):
-        port = netloc.split(':', 1)[1]
-        if not port.isdigit():
-            raise RuntimeError("%r is not a valid port number." % port)
+        host, port = ["0.0.0.0", default_port]
```
```py
host, port = "0.0.0.0", default_port
```

berkerpeksag commented on gunicorn/util.py:
```diff
@@ -232,33 +232,30 @@ def is_ipv6(addr):
-        port = default_port
-    return (host, port)
+    except ValueError:
+        raise RuntimeError("%r is not a valid port number." % port)
```
Since master now targets Gunicorn 20, perhaps we can raise `ValueError` as it's more correct exception to raise. @benoitc @tilgovi what do you think?

```py
except ValueError:
    raise ValueError("%r is not a valid port number." % port) from None
```

tilgovi replied:
Sounds good to me!

berkerpeksag commented on gunicorn/util.py:
```diff
@@ -232,33 +232,30 @@ def is_ipv6(addr):
+    except ValueError:
+        raise RuntimeError("%r is not a valid port number." % port)
+
+    return (host.lower(), port)
```
Style nit:

```py
return host.lower(), port
```

hramezani commented:
@berkerpeksag thanks, I fixed all of your comments. Also rebase the PR with master

benoitc commented on tests/test_util.py:
```diff
@@ -24,7 +27,7 @@ def test_parse_address(test_input, expected):
 
 def test_parse_address_invalid():
-    with pytest.raises(RuntimeError) as err:
+    with pytest.raises(ValueError) as err:
```
RuntimeError should be kept there as it used by the CLI.

benoitc commented on gunicorn/util.py:
```diff
@@ -247,33 +247,28 @@ def is_ipv6(addr):
-        host = netloc.lower()
-
-    #get port
-    netloc = netloc.split(']')[-1]
```
but then what if a port different from the default  is given via the netloc ?

benoitc commented on tests/test_util.py:
```diff
@@ -13,8 +13,11 @@
 @pytest.mark.parametrize('test_input, expected', [
```
tests are missing the case where the port in the netloc string is different from the default.

hramezani commented:
@benoitc thanks for the comment. I changed the `ValueError` to `RuntimeError`.
Also, I added a test for the case which port in `netloc`  is different from the default.

tilgovi commented:
I think we should merge this as-is. If we would like to change the `RuntimeError` on master, we can do that separately. The patch is good the way it is, though, and doesn't change the exception behavior.

tilgovi commented:
Quiet for a while now and everything looks good. Merging! Thanks for your help and patience, @hramezani.
